fairlearn/metrics/_disaggregated_metric.py

```diff
@@ -0,0 +1,301 @@
+# Copyright (c) Microsoft Corporation and Fairlearn contributors.
+# Licensed under the MIT License.
+
+import logging
+from typing import Dict, List, Optional, Union
+
+import numpy as np
+import pandas as pd
+
+from ._annotated_metric_function import AnnotatedMetricFunction
+
+logger = logging.getLogger(__name__)
+
+_VALID_ERROR_STRING = ["raise", "coerce"]
+_VALID_GROUPING_FUNCTION = ["min", "max"]
+
+_INVALID_ERRORS_VALUE_ERROR_MESSAGE = (
+    "Invalid error value specified. Valid values are {0}".format(_VALID_ERROR_STRING)
+)
+_INVALID_GROUPING_FUNCTION_ERROR_MESSAGE = (
+    "Invalid grouping function specified. Valid values are {0}".format(
+        _VALID_GROUPING_FUNCTION
+    )
+)
+_MF_CONTAINS_NON_SCALAR_ERROR_MESSAGE = (
+    "Metric frame contains non-scalar cells. Please remove non-scalar columns from your"
+    " metric frame or use parameter errors='coerce'."
+)
+
+
+def extract_unique_classes(
+    data: pd.DataFrame, feature_list: List[str]
+) -> Dict[str, np.ndarray]:
+    """Compute unique values in a given set of columns."""
+    result = dict()
+    for feature in feature_list:
+        result[feature] = np.unique(data[feature])
+    return result
+
+
+def apply_to_dataframe(
+    data: pd.DataFrame, metric_functions: Dict[str, AnnotatedMetricFunction]
+) -> pd.Series:
+    """Apply metric functions to a DataFrame.
+
+    The incoming DataFrame may have been sliced via `groupby()`.
+    This function applies each annotated function in turn to the
+    supplied DataFrame.
+    """
+    values = dict()
+    for function_name, metric_function in metric_functions.items():
+        values[function_name] = metric_function(data)
+    result = pd.Series(data=values.values(), index=values.keys())
+    return result
+
+
+class DisaggregatedResult:
+    """Pickier version of MetricFrame.
+
+    This holds the internal result from a disaggregated metric
+    computation, and provides `apply_grouping()` (to cover min
+    and max), `difference()` and `ratio()` methods.
+
+    The main difference to the results computed by MetricFrame
+    is that no account is made of whether the user supplied
+    a bare function or a dictionary. Hence the results are
+    always Series or DataFrame
+    """
```

Comment:
would be nice to have the `Parameters` section here for the inputs and hat they are supposed to be.

Reply:
Done